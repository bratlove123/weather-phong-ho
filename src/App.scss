.app-container {
  background-color: #f6f6f6;
  position: absolute;

  width: 1200px;
  height: 600px;

  /* Center form on page horizontally & vertically */
  top: 50%;
  left: 50%;
  margin-top: -300px;
  margin-left: -600px;

  box-shadow: 0 10px 20px rgba(0,0,0,0.19), 0 6px 6px rgba(0,0,0,0.23);

  border-radius: 20px;

  overflow-y: auto;
}
