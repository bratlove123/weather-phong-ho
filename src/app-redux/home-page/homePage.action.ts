import { WeatherResponse } from "models";

export const GET_DATES = "GET_DATES";
export const SET_LOADING = "SET_LOADING";
export const SEARCH_LOCATIONS = "SEARCH_LOCATIONS";
export const GET_SEARCH_LOCATIONS_SUCCESS = "SEARCH_LOCATIONS_SUCCESS";
export const GET_SEARCH_LOCATIONS_ERROR = "GET_SEARCH_LOCATIONS_ERROR";

interface GetDatesAction {
  type: typeof GET_DATES;
}

interface SetLoadingAction {
  type: typeof SET_LOADING;
  payload: boolean;
}

export interface SearchLocations {
  type: typeof SEARCH_LOCATIONS;
  payload: string;
}

export interface GetSearchLocationsSuccess {
  type: typeof GET_SEARCH_LOCATIONS_SUCCESS;
  payload: WeatherResponse;
}

export interface GetSearchLocationsError {
  type: typeof GET_SEARCH_LOCATIONS_ERROR;
  payload: string;
}

export type HomePageTypes =
  | GetDatesAction
  | SetLoadingAction
  | GetSearchLocationsError
  | SearchLocations
  | GetSearchLocationsSuccess;
