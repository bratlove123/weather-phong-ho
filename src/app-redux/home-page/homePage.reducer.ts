import produce from "immer";
import { HomePageState, WeatherByDate } from "models";
import {
  GET_SEARCH_LOCATIONS_ERROR,
  GET_SEARCH_LOCATIONS_SUCCESS,
  HomePageTypes,
  SEARCH_LOCATIONS,
  SET_LOADING,
} from "./homePage.action";

// Define your state here
const initialState: HomePageState = {
  isLoading: false,
  dates: [],
  error: "",
  cityId: NaN,
};

export const homePage = (state = initialState, action: HomePageTypes) =>
  produce(state, (draftState) => {
    switch (action.type) {
      case SET_LOADING:
        draftState.isLoading = action.payload;
        break;
      case SEARCH_LOCATIONS:
        draftState.isLoading = true;
        draftState.dates = [];
        draftState.error = "";
        break;
      case GET_SEARCH_LOCATIONS_ERROR:
        draftState.error = action.payload;
        draftState.isLoading = false;
        break;
      case GET_SEARCH_LOCATIONS_SUCCESS:
        let dates = action.payload.list;
        let datesFilter: WeatherByDate[] = [];
        dates.forEach((date) => {
          if (
            datesFilter.length === 0 ||
            (datesFilter.length > 0 &&
              datesFilter.length < 5 &&
              new Date(date.dt * 1000).setHours(0, 0, 0, 0) >
                new Date(
                  datesFilter[datesFilter.length - 1].dt * 1000
                ).setHours(0, 0, 0, 0))
          ) {
            datesFilter.push(date);
          }
        });

        draftState.dates = datesFilter;
        draftState.isLoading = false;
        draftState.cityId = action.payload.city.id
        break;
      default:
        return draftState;
    }
  });
