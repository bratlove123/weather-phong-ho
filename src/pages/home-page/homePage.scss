.home-page-container {
  display: flex;
  flex-direction: column;

  .search-container {
    padding: 20px;
    display: flex;

    .search {
      margin: auto;
    }
  }

  .weather-container {
    padding: 40px;
  }

  .title {
    font-size: 50px;
    text-align: center;
    padding: 30px;
    font-weight: bold;
    color: #407294;
    font-family: monospace;
  }

  .error {
    font-size: 25px;
    color: #e64857;
    margin: 70px;
    text-transform: uppercase;
    text-align: center;
    font-weight: bold;
  }
}
