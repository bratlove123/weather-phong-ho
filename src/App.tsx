import React from "react";
import { Provider } from "react-redux";
import store from "app-redux/store";
import "./App.scss";
import { HomePage } from "pages/home-page/homePage";

function App() {
  return (
    <Provider store={store}>
      <div className="app-container">
        <HomePage></HomePage>
      </div>
    </Provider>
  );
}

export default App;
