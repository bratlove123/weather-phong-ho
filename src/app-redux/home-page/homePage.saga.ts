import { WeatherResponse } from "models";
import { call, put, takeLatest } from "redux-saga/effects";
import { api } from "../../services/homePage.service";

// Import all actions
import {
  SearchLocations,
  SEARCH_LOCATIONS,
  GET_SEARCH_LOCATIONS_SUCCESS,
  GET_SEARCH_LOCATIONS_ERROR,
} from "./homePage.action";

export function* searchLocations(action: SearchLocations) {
  const locationsResponse: WeatherResponse = yield call(
    api.searchLocationsRequest,
    action.payload
  );

  if (locationsResponse && locationsResponse.cod === "200") {
    yield put({
      type: GET_SEARCH_LOCATIONS_SUCCESS,
      payload: locationsResponse,
    });
  } else {
    yield put({
      type: GET_SEARCH_LOCATIONS_ERROR,
      payload: locationsResponse?.message.indexOf("404")
        ? "City not found"
        : locationsResponse?.message ?? "Service error.",
    });
  }
}

export default function* homePageSaga() {
  yield takeLatest<SearchLocations>(SEARCH_LOCATIONS, searchLocations);
}
