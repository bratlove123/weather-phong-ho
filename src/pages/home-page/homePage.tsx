import { SEARCH_LOCATIONS } from "app-redux";
import { RootState } from "models/states/root-state";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { SearchInput } from "shared-components/inputs/search-input/searchInput";
import { DatesWeather } from "./components/dates-weather/datesWeather";
import { Loading } from "./components/loading-weather/loadingWeather";

import "./homePage.scss";

const isLoadingState = (state: RootState) => state.homePage.isLoading;
const errorMessageState = (state: RootState) => state.homePage.error;

export const HomePage = () => {
  const isLoading = useSelector(isLoadingState);
  const errorMessage = useSelector(errorMessageState);
  const dispatch = useDispatch();

  return (
    <div className="home-page-container">
      <div className="title">WEATHER FORECAST</div>
      <div className="search-container">
        <div className="search">
          <SearchInput
            placeholder="Type a city. Example: London"
            saving={isLoading}
            onSubmitSearch={(val: string) => {
              dispatch({ type: SEARCH_LOCATIONS, payload: val });
            }}
          ></SearchInput>
        </div>
      </div>
      <div className="weather-container">
        {isLoading && <Loading />}
        {errorMessage !== "" && (
          <div className="error">Opps! {errorMessage}</div>
        )}
        <DatesWeather></DatesWeather>
      </div>
    </div>
  );
};
