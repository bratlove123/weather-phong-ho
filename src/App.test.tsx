import React from "react";
import App from "./App";
import { shallow, mount } from "enzyme";
import { HomePage } from "pages/home-page/homePage";

test("renders the component", () => {
  const component = shallow(<App />);
  expect(component).toMatchSnapshot();
});

test("app contains home page", () => {
  const wrapper = mount(<App />);
  expect(wrapper.contains(<HomePage></HomePage>)).toBeTruthy();
});