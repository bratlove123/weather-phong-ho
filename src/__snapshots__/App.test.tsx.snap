// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<Fragment>
  <header />
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <div
      className="app-container"
    >
      <Component />
    </div>
  </Provider>
</Fragment>
`;
